ring/util.go: make WaitRingStability more responsive

krajorama opened the pull request:
**What this PR does**:

In some use cases, such as the Cortex store gateway, it is enough for the
user to know that the token distribution didn't change and it is not
important to know if some member changed between allowed states.

This patch implements the new WaitRingTokensStability function to
be able to wait for tokens stability and ignore allowed state changes.

The original motivation was that store gateway tests took a minute longer
then strictly necessary.

**Which issue(s) this PR fixes**:

Fixes #96 

**Checklist**
- [+] Tests updated
- [+] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

aknuds1 commented on CHANGELOG.md:
```diff
@@ -24,3 +24,4 @@
 * [ENHANCEMENT] flagext: for cases such as `DeprecatedFlag()` that need a logger, add RegisterFlagsWithLogger. #80
 * [BUGFIX] spanlogger: Support multiple tenant IDs. #59
 * [BUGFIX] Memberlist: fixed corrupted packets when sending compound messages with more than 255 messages or messages bigger than 64KB. #85
+* [ENHANCEMENT] ring/util.go: make WaitRingStability more responsive #96
```
This PR's number is 95:
```suggestion
* [ENHANCEMENT] ring/util.go: make WaitRingStability more responsive #95
```

krajorama replied:
mixup, thought it was the issue to be referenced

aknuds1 commented on ring/util.go:
```diff
@@ -102,27 +102,28 @@ func WaitRingStability(ctx context.Context, r *Ring, op Operation, minStability,
+	// This timer will be our state
+	stabilityTimer := time.NewTimer(minStability)
+
+	stateWatch, _ := r.statesWatcher.Register()
```
Why ignore the error?

krajorama replied:
fair

aknuds1 commented on ring/util.go:
```diff
@@ -102,27 +102,28 @@ func WaitRingStability(ctx context.Context, r *Ring, op Operation, minStability,
 		case <-ctx.Done():
 			return ctx.Err()
-		case <-pollingTicker.C:
-			// We ignore the error because in case of error it will return an empty
-			// replication set which we use to compare with the previous state.
```
Why remove this comment, isn't it still relevant?

krajorama replied:
heavy handed, removed the whole case and rebuilt it. fixed

aknuds1 commented on ring/watcher.go:
```diff
@@ -0,0 +1,54 @@
+	"sync"
+)
+
+type Watcher interface {
```
Why do we need this public interface in the package? Can it just be removed since it's not in use?

krajorama replied:
okok, I almost used it for something else as well, but it did not improve speed

aknuds1 commented on ring/watcher.go:
```diff
@@ -0,0 +1,54 @@
+}
+
+// Handles notifying listeners about state change in the Ring
+// Assumes that the callee will check that there is meaningfull change in the states
```
I think the correct term would be "the caller", i.e. the caller of the method:
```suggestion
// Assumes that the caller will check that there is meaningful change to the state.
```

krajorama replied:
depends on how you look at it I suppose. I'll change it to reflect the user experience more

aknuds1 replied:
The callee would be this method, wouldn't it since it's getting called? It's the caller of the method that should ensure there's meaningful state change right?

aknuds1 commented on ring/watcher.go:
```diff
@@ -0,0 +1,54 @@
+
+type watcher struct {
+	mtx      sync.Mutex
+	notifier chan interface{}
```
Isn't it more readable if the type of the channel is `struct{}`, i.e. signaling that no information is passed over the channel?
```suggestion
	notifier chan struct{}
```

krajorama replied:
I'm just a noob at golang, fixed

aknuds1 reviewed (commented):
I left some questions, please have a look :)

krajorama commented:
@aknuds1 Thank you for your comments! What do you think about this solution instead of reducing the polling interval overall?

aknuds1 commented:
> @aknuds1 Thank you for your comments! What do you think about this solution instead of reducing the polling interval overall?

Happy to help :) I don't know enough about the current solution yet to tell if this is a good idea or not, so it's better if someone like Peter or Marco f.ex. comments. I have some other obligations now, but I hope to return to take another pass at reviewing this PR.

pracucci commented on ring/util.go:
```diff
@@ -102,27 +102,37 @@ func WaitRingStability(ctx context.Context, r *Ring, op Operation, minStability,
+			if !stabilityTimer.Stop() {
+				<-stabilityTimer.C
+			}
+			stabilityTimer.Reset(minStability)
```
Here we reset the timer but the fact that we received a notification from the watch doesn't mean the ring has changed so that `HasReplicationSetChanged()` return true (for example, that function doesn't compare heartbeat `Timestamp` which change continuously but the watcher will notify whenever an heartbeat timestamp has changed too).

pracucci reviewed (commented):
Thanks for working on this! You raised a very good issue, however I think the offered solution is not the right one. Let me explain it.

This solution assumes that ring changes propagation are almost instantaneous between different members but that's not guaranteed. We have relatively short propagation times when running ring on Consul (but still in the order of 1s or 2s) and higher propagation times when running with memberlist (even few tens of seconds on very large clusters). If propagation times are not almost instantaneous, the issue you describe happen anyway.

After looking at this along with Peter and after looking at how we use this in Cortex, we think what we really want is "wait ring TOKENS stability". Our suggestion is:
1. Add `WaitRingTokensStability()` which is a copy of `WaitRingStability()`
2. In `WaitRingTokensStability()` instead of calling `HasReplicationSetChanged()`, call a different function which is a copy of `HasReplicationSetChanged()` but doesn't compare member states (as well as doesn't compare timestamps too as `HasReplicationSetChanged()`)

_In the suggestion above I mentioned "copy" to make it clear what we have in mind, but we can obviously refactor existing methods to avoid code duplication. Just keep in mind we shouldn't change behaviour and signature of `HasReplicationSetChanged()` and `WaitRingStability()` in order to not introduce breaking changes._

## Why this PR assumes almost instantaneous propagation times?

This about a simple setup with 2 members (A and B) and the following timeline:
- A and B wait for stability
- A reaches stability
- A moves from JOINING -> ACTIVE
- B receives all updates at once and, despite the ring stability was reached, it resets the counter because A reached the stability but already changed their state from JOINING to ACTIVE (and what we're trying to fix is that B shouldn't reset the timer in this case)

With memberlist it's even worse because there's no changes propagation ordering guaranteed (while with Consul CAS() changes are propagated in order, but multiple changes can be propagated together).

aknuds1 commented:
Thanks @pracucci for chiming in! :)

krajorama commented:
I've updated the solution with the suggested way of working. Added new tests and tested cortex as well - the expected speedup in the tests is still there.

pracucci commented on CHANGELOG.md:
```diff
@@ -24,3 +24,4 @@
 * [ENHANCEMENT] flagext: for cases such as `DeprecatedFlag()` that need a logger, add RegisterFlagsWithLogger. #80
 * [BUGFIX] spanlogger: Support multiple tenant IDs. #59
 * [BUGFIX] Memberlist: fixed corrupted packets when sending compound messages with more than 255 messages or messages bigger than 64KB. #85
+* [ENHANCEMENT] Add WaitRingTokensStability function to ring, to be able to wait on ring stability excluding allowed state transitions. #95
```
[nit] We're used to group changelog entries by type. Do you mind moving this up, grouping to the other `[ENHANCEMENT]` entries, please?

krajorama replied:
ack

pracucci commented on ring/util.go:
```diff
@@ -96,6 +96,19 @@ func WaitInstanceState(ctx context.Context, r ReadRing, instanceID string, state
+}
+
+// WaitRingTokensStability waits for the Ring to be unchanged at
+// least for minStability time period, excluding transitioninig between
```
[nit]
```suggestion
// least for minStability time period, excluding transitioning between
```

krajorama replied:
thank you, I didn't realize that VS code has no spell check by default like Rider. Now installed.

pracucci commented on ring/replication_set_test.go:
```diff
@@ -241,3 +241,60 @@ func TestReplicationSet_Do(t *testing.T) {
+
+func TestHasReplicationSetChangedWithoutStateIgnoresTimeStampAndState(t *testing.T) {
+	// Only testing difference to underlying Equal function
+	rs1 := ReplicationSet{
```
[nit] May calling this `initialState` (or something like this) may clarify this is the initial state we're comparing to?

krajorama replied:
fair

pracucci commented on ring/replication_set_test.go:
```diff
@@ -241,3 +241,60 @@ func TestReplicationSet_Do(t *testing.T) {
+		},
+	}
+	tests := map[string]struct {
+		rs                                         ReplicationSet
```
[nit] Maybe calling it `nextState` could help to clarify it. Also naming fields in the test case. They're nit, up to you whether picking it up or not.

krajorama replied:
fair

pracucci commented on ring/util_test.go:
```diff
@@ -123,48 +116,50 @@ func TestWaitRingStabilityShouldReturnAsSoonAsMinStabilityIsReachedOnNoChanges(t
+	t.Parallel()
+
+	const (
+		minStability = 2 * time.Second
```
2s is not enough here, because when we assert on the actual wait time we pass a delta of 2s, see:
```
assert.InDelta(t, minStability, elapsedTime, float64(2*time.Second))
```

If we remove the call to `WaitRingStability()` at all, the test will pass anyway. I know it's not something you changed, but I would be glad if you could fix it. One way to fix it is checking the elapsed time like we do in `TestWaitRingStabilityShouldReturnOnceMinStabilityOfInstancesHasBeenReached`.

_The same comment applies to tests below._

krajorama replied:
Yes, I noticed this as well,  I'll add the checking for minimal time. I suspect the numbers here are chosen to avoid flakyness.  
I also think that these should not be running in real time, so I'll propose a change as part of my pet project to improve testability at a later time.

pracucci commented on ring/util_test.go:
```diff
@@ -240,6 +261,98 @@ func TestWaitRingStabilityShouldReturnErrorIfMaxWaitingIsReached(t *testing.T) {
 	assert.InDelta(t, maxWaiting, elapsedTime, float64(2*time.Second))
```
[nit] I think we can assert on elapsedTime >= maxWaiting instead of using delta, because we expect at least maxWaiting time has been waited for.

_Same comment applies to tests below._

krajorama replied:
yes

pracucci commented on ring/util_test.go:
```diff
@@ -240,6 +261,98 @@ func TestWaitRingStabilityShouldReturnErrorIfMaxWaitingIsReached(t *testing.T) {
+	require.Equal(t, nil, WaitRingTokensStability(context.Background(), ring, Reporting, minStability, maxWaiting))
+	elapsedTime := time.Since(startTime)
+
+	assert.InDelta(t, minStability, elapsedTime, float64(2*time.Second))
```
Same comment here. If we skip calling `WaitRingTokensStability()` at all the test will pass anyway.

pracucci reviewed (approved):
Thanks for working on this! Good job. The refactoring is very clean. The logic changes LGTM, but I would just improve a bit test assertions before merging. Thanks!

krajorama commented on ring/replication_set_test.go:
```diff
@@ -241,3 +241,60 @@ func TestReplicationSet_Do(t *testing.T) {
 	}
 }
+
+func TestHasReplicationSetChangedWithoutStateIgnoresTimeStampAndState(t *testing.T) {
```
fixed

krajorama commented on ring/replication_set_test.go:
```diff
@@ -241,3 +241,60 @@ func TestReplicationSet_Do(t *testing.T) {
+	}
+	for testName, testData := range tests {
+		t.Run(testName, func(t *testing.T) {
+			assert.Equal(t, testData.expectHasReplicationSetChanged, HasReplicationSetChanged(rs1, testData.rs), "HasReplicationSetChanged wrong result")
```
I'm trying to not repeat a lot of code and point out the difference between the two functions. That said, the naming was crap. Let me suggest a new way

krajorama commented on ring/util_test.go:
```diff
@@ -123,48 +116,50 @@ func TestWaitRingStabilityShouldReturnAsSoonAsMinStabilityIsReachedOnNoChanges(t
+	return ring
+}
+
+func TestWaitRingStabilityShouldReturnAsSoonAsMinStabilityIsReachedOnNoChanges(t *testing.T) {
```
fixed

pracucci reviewed (approved):
LGTM, thanks! 👏

pstibrany commented on ring/util_test.go:
```diff
@@ -96,21 +96,14 @@ func TestGenerateTokensIgnoresOldTokens(t *testing.T) {
-		"instance-3": {Addr: "127.0.0.3", State: JOINING, Timestamp: time.Now().Unix()},
-		"instance-4": {Addr: "127.0.0.4", State: LEAVING, Timestamp: time.Now().Unix()},
-		"instance-5": {Addr: "127.0.0.5", State: ACTIVE, Timestamp: time.Now().Unix()},
+		"instance-1": {Addr: "instance-1", State: ACTIVE, Timestamp: time.Now().Unix()},
```
In practice, this field contains an IP address, so previous example was bit better, although in this test, it doesn't matter.

krajorama replied:
fixed

pstibrany commented on ring/util_test.go:
```diff
@@ -182,44 +194,40 @@ func TestWaitRingStabilityShouldReturnOnceMinStabilityHasBeenReached(t *testing.
-		strategy:            NewDefaultReplicationStrategy(),
-	}
+	// Add 1 new instance after some time.
+	go addInstanceAfterSomeTime(ring, addInstanceAfter)
```
This function already starts extra goroutine, so this `go` keyword is currently extra.

pstibrany commented on ring/util_test.go:
```diff
@@ -232,15 +240,123 @@ func TestWaitRingStabilityShouldReturnErrorIfMaxWaitingIsReached(t *testing.T) {
+
+	// Keep changing the ring.
+	done := changeStatePeriodically(ring)
+	defer close(done)
```
Nit: This test leaves goroutine running even after test is finishes (for up to 1s). Other tests  have the same or similar problems (eg. leaving goroutines behind, if they fail early). Goleak took would flag such tests, even though goroutines finish soon after. Probably not worth fixing.

krajorama replied:
huh? I'm confused, the done channel is there specifically to avoid that. How would it leak?

pstibrany replied:
Oh, somehow I read the code as if it was waiting 1s. My bad. The race between test end and goroutine exit is still there, but it's definitely not 1s.

krajorama replied:
discussed offline, won't fix now

pstibrany commented on ring/util_test.go:
```diff
@@ -232,15 +240,123 @@ func TestWaitRingStabilityShouldReturnErrorIfMaxWaitingIsReached(t *testing.T) {
+	elapsedTime := time.Since(startTime)
+
+	assert.GreaterOrEqual(t, elapsedTime.Milliseconds(), minStability.Milliseconds())
+	assert.Less(t, elapsedTime.Milliseconds(), 2*minStability.Milliseconds())
```
```suggestion
	assert.Less(t, elapsedTime, 2*minStability)
```

(Everywhere, not just here.)

krajorama replied:
lol , true, don't know why it was like that in the first place, removed

pstibrany reviewed (approved):
Changes look good. I left some non-blocking comments.

aknuds1 commented on ring/util_test.go:
```diff
@@ -232,15 +240,123 @@ func TestWaitRingStabilityShouldReturnErrorIfMaxWaitingIsReached(t *testing.T) {
+	defer close(done)
+
+	startTime := time.Now()
+	require.Equal(t, nil, WaitRingTokensStability(context.Background(), ring, Reporting, minStability, maxWaiting))
```
Nit:
```suggestion
	require.NoError(t, WaitRingTokensStability(context.Background(), ring, Reporting, minStability, maxWaiting))
```

aknuds1 reviewed (approved):
LGTM, just a nit.

pstibrany commented on ring/util_test.go:
```diff
@@ -257,10 +257,10 @@ func TestWaitRingStability_ShouldReturnErrorIfInstancesAddedAndMaxWaitingIsReach
 
 	startTime := time.Now()
-	require.Equal(t, context.DeadlineExceeded, WaitRingStability(context.Background(), ring, Reporting, minStability, maxWaiting))
+	require.Error(t, context.DeadlineExceeded, WaitRingStability(context.Background(), ring, Reporting, minStability, maxWaiting))
```
This currently tests that `context.DeadlineExceeded` is an error, not that `WaitRingStability` returned one.

pstibrany replied:
Btw, Same problem is in other places of this PR (eg. lines 280, 331)

krajorama replied:
dang, fixed
